syntax = "proto3";

service TrackerServerService {
    rpc AskFileOperation(AskFileOperationRequest) returns (AskFileOperationResponse) {}
    rpc GetFileList(GetFileListRequest) returns (GetFileListResponse) {}
    rpc UpdateFileInfo(UpdateFileInfoRequest) returns (UpdateFileInfoResponse) {}
    rpc Replicate(ReplicateRequest) returns (ReplicateResponse) {}
    rpc AddStorageServer(AddStorageServerRequest) returns (AddStorageServerResponse) {}
    rpc DeleteFile(DeleteFileRequest) returns (DeleteFileResponse) {}
}

message AskFileOperationRequest {
    string operation = 1; 
    string filename = 2;
    string group_id = 3;
}

message AskFileOperationResponse {
    string status = 1;
    string ip = 2;
    string port = 3;
}

message GetFileListRequest {
    string operation = 1; // null
}

message GetFileListResponse {
    message Group {
        message File {
            string file_name = 1;
            string file_path = 2;
        }
        int32 group_id = 1;
        repeated File files = 2;        
    }
    repeated Group groups = 1;
}

message UpdateFileInfoRequest {
    string filename = 1;
    string group_id = 2;
    string filepath = 3;
}

message UpdateFileInfoResponse {
    string status = 1;
    string time = 2;
}

message ReplicateRequest {
    string port = 1;
    string ip = 2;
}

message ReplicateResponse {
    // return the message with same group
    bool status = 1;
    string port = 2;
    string ip = 3;
}

message AddStorageServerRequest {
    string group_id = 1;
    string ip = 2;
    string port = 3;
    string storage_server_id = 4;
}

message AddStorageServerResponse {
    string status = 1;
}

message DeleteFileRequest {
    string filename = 1;
    string filepath = 2;
    string group_id = 3;
}

message DeleteFileResponse {
    string status = 1;
}